jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/GraphIndexWriter.java

```diff
@@ -38,4 +53,62 @@ public interface GraphIndexWriter extends Closeable {
      * @param featureStateSuppliers a map of FeatureId to a function that returns a Feature.State
      */
     void write(Map<FeatureId, IntFunction<Feature.State>> featureStateSuppliers) throws IOException;
+
+    /**
+     * Factory method to obtain a builder for the specified writer type with an IndexWriter.
+     * <p>
+     * This overload accepts any IndexWriter but certain types have specific requirements:
+     * <ul>
+     *   <li>ON_DISK requires a RandomAccessWriter (will throw IllegalArgumentException otherwise)</li>
+     *   <li>ON_DISK_SEQUENTIAL accepts any IndexWriter</li>
+     *   <li>ON_DISK_PARALLEL is not supported via this method (use the Path overload instead)</li>
+     * </ul>
+     *
+     * @param type the type of writer to create
+     * @param graphIndex the graph index to write
+     * @param out the output writer
+     * @return a builder for the specified writer type
+     * @throws IllegalArgumentException if the type requires a specific writer type that wasn't provided
+     */
+    static AbstractGraphIndexWriter.Builder<? extends AbstractGraphIndexWriter<?>, ? extends IndexWriter>
```

Comment:
Would it be useful to separate AbstractGraphIndexWriter.Builder into a separate file so that we decouple the implementations from the interface in the code?